mm2src/mm2_main/src/lp_swap/swap_watcher.rs

```diff
@@ -1,32 +1,63 @@
-use super::{broadcast_p2p_tx_msg, lp_coinfind, tx_helper_topic, wait_for_taker_payment_conf_until, H256Json,
-            SwapsContext, TransactionIdentifier, WAIT_CONFIRM_INTERVAL};
+use super::{broadcast_p2p_tx_msg, lp_coinfind, taker_payment_spend_deadline, tx_helper_topic, H256Json, SwapsContext,
+            WAIT_CONFIRM_INTERVAL};
 use crate::mm2::MmError;
 use async_trait::async_trait;
-use coins::{CanRefundHtlc, MmCoinEnum, WatcherValidatePaymentInput};
+use coins::{CanRefundHtlc, FoundSwapTxSpend, MmCoinEnum, WatcherSearchForSwapTxSpendInput,
+            WatcherValidatePaymentInput, WatcherValidateTakerFeeInput};
 use common::executor::{AbortSettings, SpawnAbortable, Timer};
-use common::log::{error, info};
-use common::now_ms;
+use common::log::{debug, error, info};
 use common::state_machine::prelude::*;
+use common::{now_ms, DEX_FEE_ADDR_RAW_PUBKEY};
 use futures::compat::Future01CompatExt;
 use mm2_core::mm_ctx::MmArc;
 use mm2_libp2p::{decode_signed, pub_sub_topic, TopicPrefix};
-use mm2_number::BigDecimal;
-use rpc::v1::types::Bytes as BytesJson;
+use serde::{Deserialize, Serialize};
+use serde_json as json;
 use std::cmp::min;
 use std::sync::Arc;
 use uuid::Uuid;
 
 pub const WATCHER_PREFIX: TopicPrefix = "swpwtchr";
 const TAKER_SWAP_CONFIRMATIONS: u64 = 1;
 pub const TAKER_SWAP_ENTRY_TIMEOUT: u64 = 21600;
-const WAIT_FOR_TAKER_REFUND: u64 = 1200; // How long?
+
+pub const MAKER_PAYMENT_SPEND_SENT_LOG: &str = "Maker payment spend sent";
+pub const MAKER_PAYMENT_SPEND_FOUND_LOG: &str = "Maker payment spend found by watcher";
+pub const TAKER_PAYMENT_REFUND_SENT_LOG: &str = "Taker payment refund sent";
 
 struct WatcherContext {
     ctx: MmArc,
     taker_coin: MmCoinEnum,
     maker_coin: MmCoinEnum,
-    data: TakerSwapWatcherData,
     verified_pub: Vec<u8>,
+    data: TakerSwapWatcherData,
+    conf: WatcherConf,
+}
+
+impl WatcherContext {
+    fn taker_locktime(&self) -> u64 { self.data.swap_started_at + self.data.lock_duration }
+
+    fn wait_for_maker_payment_spend_deadline(&self) -> u64 {
+        let factor = self.conf.wait_maker_payment_spend_factor;
+        self.data.swap_started_at + (factor * self.data.lock_duration as f64) as u64
+    }
+
+    fn refund_start_time(&self) -> u64 {
+        let factor = self.conf.refund_start_factor;
+        self.data.swap_started_at + (factor * self.data.lock_duration as f64) as u64
+    }
+}
+
+#[derive(Serialize, Deserialize, Default)]
```

Comment:
With derived `Default`, we get the following:
```rust
    let conf = WatcherConf::default();
    println!("{:?}", conf);
```
`
WatcherConf { wait_taker_payment: 0.0, wait_maker_payment_spend_factor: 0.0, refund_start_factor: 0.0, search_interval: 0.0 }
`

I suppose it's not what we expect to get here:
https://github.com/KomodoPlatform/atomicDEX-API/blob/dec9c70563aaa84cae8ce8d119643ed6fe0b8067/mm2src/mm2_main/src/lp_swap/swap_watcher.rs#L580    

Could you please implement `Default` manually instead, using appropriate values?

Reply:
I believe this is resolved